### PR TITLE
hotplug rng devices in loop

### PR DESCRIPTION
**Pull request by suqinhuang**
hotplug/unplug rng devices 500 times
and run random.exe before/after hotplug 

Signed-off-by: Suqin Huang <shuang@redhat.com>

### CODE REVIEW

**Comment by suqinhuang** [comment]
id: 1231120

**Comment by ngu-niny** [comment]
Since there exists product bz 1348044, the patch might need update.
However, even I added to execute cmd 'systemctl stop rngd' inside the guest before hot unplug the rng device, it still got 'TestFail: Still get q'virtio-rng-pci-0' in qtree after unplug' when tried to hot unplug the 1st rng device.

**Comment by xutian** [comment]
@suqinhuang  Please check ngu's comments, it looks still have bug in your patch.

**Comment by suqinhuang** [comment]
@ngu-niny @xutian patch updated, could you have a look

File: qemu/tests/rng_hotplug.py [hunk 1]
```diff
@@ -38,32 +103,17 @@ def run(test, params, env):
...
                                        {'id': 'virtio-rng-pci-%d' % num})
+            hotplug_rng(vm, new_dev)
             dev_list.append(new_dev)
-            error_context.context("Hotplug %s" % new_dev, logging.info)
-            output = new_dev.hotplug(vm.monitor)
-            time.sleep(2)
 
-            error_context.context("Check %sfrom qtree after hotplug" % new_dev,
+        # Run test after hotplug
+        if test_after_hotplug and i == xrange(repeat_times):
```
Comment (pingl2016) [inline qemu/tests/rng_hotplug.py]: should be "i == repeat_times-1"?
\>>> repeat_times = 10
\>>> xrange(repeat_times)
xrange(10)
\>>> type(xrange(repeat_times))
\<type 'xrange'\>

**Comment by ngu-niny** [comment]
Test on latest rhev7.3 host: Manually without any issue; while auto for rng-egd always failed for 'Still get RNG_DEVICE in qtree after unplug'.
Test on latest rhel7.3 and rhev7.2z hosts: Both manually and auto have some issues; the auto always failed for 'Still get RNG_DEVICE in qtree after unplug'.

For the issues existed in manually test of rhev7.2z and rhel7.3, will check with the developer Amit if available; however, as bz1348044 showed, the virtio rng hotunplug test is in low priority.

**Comment by suqinhuang** [comment]
did you attach device on the cmd ?

**Comment by suqinhuang** [comment]
@pingl2016 patch updated, could you have a look

File: qemu/tests/cfg/rng_hotplug.cfg [hunk 2]
```diff
@@ -4,18 +4,33 @@
...
-    sub_test_after_hotplug = rng_bat
+    rng_num = 1
     read_rng_timeout = 360
+    no no_virtio_rng
     Windows:
         session_cmd_timeout = 240
         read_rng_cmd  = "X:\random_%PROCESSOR_ARCHITECTURE%.exe"
         driver_name = "viorng"
         rng_data_rex = "0x\w"
+        driver_id_cmd = C:\devcon.exe find * | find "VirtIO"
```
Comment (pingl2016) [inline qemu/tests/cfg/rng_hotplug.cfg]: It should be write as below.
driver_id_cmd = X:\devcon\wxp_x86\devcon.exe find * | find "VirtIO"
driver_check_cmd = X:\devcon\wxp_x86\devcon.exe status @DRIVER_ID

File: qemu/tests/rng_hotplug.py [hunk 3]
```diff
@@ -23,47 +23,102 @@ def run(test, params, env):
...
-                msg += "the %sth time" % (new_dev, i)
-                raise exceptions.TestFail(msg)
-            logging.info("virtio-rng-pci-%d is hotpluged successfully" % num)
-        sub_test = params.get("sub_test_after_hotplug")
-        if sub_test:
-            utils_test.run_virt_sub_test(test, params, env, sub_test)
+            utils_test.run_virt_sub_test(test, params, env,
+                                         test_after_hotplug)
 
+        stop_rngd(vm)
```
Comment (pingl2016) [inline qemu/tests/rng_hotplug.py]: Shall we still add sleep here like that you added in line 98.

Comment (suqinhuang) [inline qemu/tests/rng_hotplug.py]: I drop the sleep

**Comment by pingl2016** [comment]
Pass the test on windows guest and failed on rhel guest as the product bug 1348044

Acked-by: Ping Li pingl@redhat.com

**Comment by ngu-niny** [comment]
LGTM, so ack it.

Acked-by: Nini Gu <ngu@redhat.com>

**Pull request merged**